### ISSUE TITLE
pl/sql used by CWMS_LOC.RETRIEVE_LOCATIONS_F doesn't escape invalid characters

### ISSUE THREAD

**Issue by rma-rripken** [USACE/cwms-data-api#259]
https://cwms-data.usace.army.mil/cwms-data/locations?names=BIGM&office=NWDM

**Comment by rma-rripken** [comment]
The result returned looks like:
{"locations":{"query-info":{"time-of-query":"2022-10-21T21:04:11Z","process-query":"PT1.209S","format-output":"PT0.000S","requested-format":"JSON","requested-office":"NWDM","requested-items":[{"name":"BIGM","unit":"EN","datum":"NATIVE"}],"total-locations-retrieved":1,"unique-locations-retrieved":1},"locations":[{"identity":{"office":"NWDM","name":"BIGM","alternate-names":[]},"label":{"public-name":null,"long-name":"Big Piney River below Ft","description":"Big Piney River below Ft. Leonard Wood, MO
"},"geolocation":{"latitude":0,"longitude":0,"horizontal-datum":null,"elevation":null},"political":{"nation":"UNITED STATES","state":null,"county":null,"timezone":null,"nearest-city":null,"bounding-office":null},"classification":{"location-kind":"SITE","location-type":null}}]}}


notice the embedded \r\n in the description

**Comment by rma-rripken** [comment]
This is the getAll end-point.  Apparently we didn't add JSONv2 to getAll.  The getOne does have JSONv2 and queries location information direct from db.  
Jackson does correctly escape/quote embedded newlines.

If we added JSONv2 to the getAll and build the Location objects similar to how its done for the getOne end-point we should be able to avoid the problem.    

If we are adding to the getAll end-point we should probably also add in paging.

**Comment by willbreitkreutz** [comment]
@MikeNeilson this one is kind of breaking how i was going to structure the demo app we're building in my course, +1 for a quick fix

**Comment by MikeNeilson** [comment]
ugg, that's not a quick fix. *BUT* it's also not the recommended format. try passing the following header:

"accept: application/json;version=2"

(that's supposed to be also doable in the query param but I'm not sure if it's in the live instance so stick with the header.)

**Comment by willbreitkreutz** [comment]
Yeah, i've been putting that header in the requests, but some of the locations calls return JSON that can't actually be parsed in the browser i.e. https://cwms-data.usace.army.mil/cwms-data/locations?office=NAE&format=json (with accept: application/json;version=2). if you look at the response in dev tools it's unable to parse.  It's not an issue with all the districts, but it seems like if anyone adds a carriage return in a text field it breaks the JSON output.

**Comment by MikeNeilson** [comment]
if the format query parameter is used it takes priority. You can use both at the same time. 
But actually I should be able to hot fix that. I need to do some paperwork first and then I'll look into it. But FYI we are trying to get people away from that particular JSON format.... as you've seen, it has... issues. Beyond just that it's rendered in the database.

**Comment by willbreitkreutz** [comment]
ok, i didn't realize those conflicted, pulled the query param out and new lines still are breaking the JSON parse.  no need to jump through hoops, just wanted to get it out there, thanks for the quick follow-up

**Comment by rma-rripken** [comment]
> (that's supposed to be also doable in the query param

Last time this came up I looked into it a little.  For JSONv2 to be settable via the queryParams I think a test like this would need to pass in FormatsTest
```
contentType = Formats.parseHeaderAndQueryParm("*/*", Formats.JSONV2);
assertNotNull(contentType);
assertEquals("application/json", contentType.getType());
parameters = contentType.getParameters();
assertNotNull(parameters);
assertFalse(parameters.isEmpty());
assertTrue(parameters.containsKey("version"));
assertEquals("2", parameters.get("version"));
```

**Comment by rma-rripken** [comment]
> ugg, that's not a quick fix. _BUT_ it's also not the recommended format. try passing the following header:
> 
> "accept: application/json;version=2"

@MikeNeilson I tried to point this out on Oct25 - the controller's getAll end-point doesn't support JSONv2.

But... @willbreitkreutz how much of the location do you need?  You might get the necessary details from the Catalog end-point. 

curl -X 'GET' \
  'https://cwms-data.usace.army.mil/cwms-data/catalog/LOCATIONS?office=NWDM&like=BIGM.%2A' \
  -H 'accept: application/json;version=2'

**Comment by MikeNeilson** [comment]
oh, yeah. thanks @rripken that's a good point. It's certainly not settled yet but we have started replacing those getAll handlers with the catalog endpoint. mostly to help simplify some of the query parameter documentation since it's all mostly the same between timeseries/locations/etc.


the catalog should also be faster. Bit of a db misconfig on the memory parameters we had to workaround for the moment.
